File: qwisp/expert_source.py
"""ExpertSource — safetensors から per-expert スライスを on-demand 取得（多シャード対応）.

格納形式（実モデル確認済）: layers.{L}.mlp.switch_mlp.{gate,up,down}_proj.{weight(U32 4bit),
scales(F16), biases(F16)}、先頭次元に 256 experts スタック → expert e は連続バイトスライス。
オフセットを safetensors ヘッダから計算し、その expert のバイトだけ pread する。
"""

from __future__ import annotations

import json
import os
import struct

import mlx.core as mx
import numpy as np

_DT = {"U32": np.uint32, "F16": np.float16, "BF16": np.uint16, "F32": np.float32}
PROJS = ("gate_proj", "up_proj", "down_proj")
PARTS = ("weight", "scales", "biases")
_PREFIX = "language_model.model.layers"


class ExpertSource:
    def __init__(self, model_dir: str, memmap: bool = False):
        self.dir = model_dir
        with open(os.path.join(model_dir, "model.safetensors.index.json")) as f:
            self.wm = json.load(f)["weight_map"]
        self._hdr: dict[str, tuple[dict, int]] = {}
        self._fd: dict[str, int] = {}
        self.use_memmap = memmap          # True で miss ロードを np.memmap fancy-index に
        self._mmap: dict[tuple, "np.memmap"] = {}

    def _memmap(self, layer: int, proj: str, part: str):
        """(layer,proj,part) の [256,...] を np.memmap で遅延 map（選択ページのみ読む）。"""
        k = (layer, proj, part)
        if k not in self._mmap:
            key = self._key(layer, proj, part)
            shard = self.wm[key]
            hdr, ds = self._header(shard)
            t = hdr[key]; b, _ = t["data_offsets"]
            self._mmap[k] = np.memmap(os.path.join(self.dir, shard), dtype=_DT[t["dtype"]],
                                      mode="r", offset=ds + b, shape=tuple(t["shape"]))
        return self._mmap[k]

    def load_expert_slices_mmap(self, layer: int, experts) -> dict:
        """memmap fancy-index で miss expert を選択読み。tensor 毎に 1 mx.array（per-slice 比 1/k 本）。"""
        idx = np.asarray(experts)
        out = {e: {} for e in experts}
        for proj in PROJS:
            for part in PARTS:
                sub = np.ascontiguousarray(self._memmap(layer, proj, part)[idx])  # [k,...] 選択
                m = mx.array(sub)
                for i, e in enumerate(experts):
                    out[e][f"{proj}.{part}"] = m[i:i + 1]
        return out

    def _key(self, layer: int, proj: str, part: str) -> str:
        return f"{_PREFIX}.{layer}.mlp.switch_mlp.{proj}.{part}"

    def _header(self, shard: str):
        if shard not in self._hdr:
            with open(os.path.join(self.dir, shard), "rb") as f:
                n = struct.unpack("<Q", f.read(8))[0]
                hdr = json.loads(f.read(n))
            self._hdr[shard] = (hdr, 8 + n)
        return self._hdr[shard]

    def _fdesc(self, shard: str) -> int:
        if shard not in self._fd:
            # v0.1: F_NOCACHE は付けない（OS ページキャッシュに hot pages を持たせる）。
            self._fd[shard] = os.open(os.path.join(self.dir, shard), os.O_RDONLY)
        return self._fd[shard]

    def slice(self, layer: int, proj: str, part: str, e: int) -> mx.array:
        key = self._key(layer, proj, part)
        shard = self.wm[key]
        hdr, data_start = self._header(shard)
        t = hdr[key]
        b, end = t["data_offsets"]
        n_exp = t["shape"][0]
        stride = (end - b) // n_exp
        buf = os.pread(self._fdesc(shard), stride, data_start + b + e * stride)
        arr = np.frombuffer(buf, _DT[t["dtype"]]).reshape([1] + t["shape"][1:])
        return mx.array(arr)

    def load_experts(self, layer: int, experts: list[int]) -> dict[str, mx.array]:
        """experts を stack（[len, ...]）で返す。proj.part キー。"""
        sub = {}
        for proj in PROJS:
            for part in PARTS:
                sub[f"{proj}.{part}"] = mx.concatenate(
                    [self.slice(layer, proj, part, e) for e in experts], axis=0)
        return sub

    def load_expert_slices(self, layer: int, experts, pool=None) -> dict:
        """miss 用: 各 expert の 9 テンソル(proj×part)を並列 pread → {e: {proj.part: mx.array}}.

        os.pread は GIL を解放するので ThreadPoolExecutor で latency を重ねられる（682 syscall/fwd 対策）。
        pool=None なら逐次。
        """
        if self.use_memmap:
            return self.load_expert_slices_mmap(layer, experts)
        out = {e: {} for e in experts}
        jobs = [(e, p, q) for e in experts for p in PROJS for q in PARTS]
        if pool is None:
            for e, p, q in jobs:
                out[e][f"{p}.{q}"] = self.slice(layer, p, q, e)
            return out
        futs = [(e, f"{p}.{q}", pool.submit(self.slice, layer, p, q, e)) for e, p, q in jobs]
        for e, k, f in futs:
            out[e][k] = f.result()
        return out

    def per_expert_bytes(self, layer: int = 0) -> int:
        total = 0
        for proj in PROJS:
            for part in PARTS:
                key = self._key(layer, proj, part)
                hdr, _ = self._header(self.wm[key])
                t = hdr[key]
                b, end = t["data_offsets"]
                total += (end - b) // t["shape"][0]
        return total

    def close(self):
        for fd in self._fd.values():
            os.close(fd)
        self._fd.clear()

File: swift/Sources/QwispCore/ExpertSource.swift
import Foundation
import MLX
#if canImport(Darwin)
import Darwin
#endif

/// safetensors から per-expert スライスを on-demand pread で取得（qwisp/expert_source.py の Swift 版）.
/// switch_mlp.{gate,up,down}_proj.{weight(U32),scales(F16),biases(F16)} は先頭次元に 256 experts
/// スタック → expert e は連続バイト。stride=(end-begin)/256、offset=dataStart+begin+e*stride を pread。
public final class ExpertSource {
    public static let projs = ["gate_proj", "up_proj", "down_proj"]
    public static let parts = ["weight", "scales", "biases"]
    static let prefix = "language_model.model.layers"

    struct TensorMeta { let dtype: String; let shape: [Int]; let begin: Int; let end: Int }

    let dir: URL
    let wm: [String: String]                          // tensor name -> shard
    var headers: [String: (meta: [String: TensorMeta], dataStart: Int)] = [:]
    var fds: [String: Int32] = [:]

    public init(modelDir: String) throws {
        dir = URL(fileURLWithPath: modelDir)
        let data = try Data(contentsOf: dir.appendingPathComponent("model.safetensors.index.json"))
        let top = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        wm = (top["weight_map"] as? [String: String]) ?? [:]
    }

    static func dtype(_ s: String) -> DType {
        switch s {
        case "U32": return .uint32
        case "F16": return .float16
        case "BF16": return .bfloat16
        case "F32": return .float32
        default: return .uint8
        }
    }

    static func itemSize(_ s: String) -> Int {
        switch s { case "U32", "F32": return 4; case "F16", "BF16": return 2; default: return 1 }
    }

    func key(_ layer: Int, _ proj: String, _ part: String) -> String {
        "\(ExpertSource.prefix).\(layer).mlp.switch_mlp.\(proj).\(part)"
    }

    func header(_ shard: String) throws -> (meta: [String: TensorMeta], dataStart: Int) {
        if let h = headers[shard] { return h }
        let url = dir.appendingPathComponent(shard)
        let fh = try FileHandle(forReadingFrom: url)
        defer { try? fh.close() }
        let lenData = try fh.read(upToCount: 8)!
        let hlen = lenData.withUnsafeBytes { $0.load(as: UInt64.self).littleEndian }
        let hdrData = try fh.read(upToCount: Int(hlen))!
        let json = try JSONSerialization.jsonObject(with: hdrData) as? [String: Any] ?? [:]
        var meta: [String: TensorMeta] = [:]
        for (k, v) in json {
            guard let d = v as? [String: Any], let dt = d["dtype"] as? String,
                  let shape = d["shape"] as? [Int],
                  let off = d["data_offsets"] as? [Int], off.count == 2 else { continue }
            meta[k] = TensorMeta(dtype: dt, shape: shape, begin: off[0], end: off[1])
        }
        let h = (meta, 8 + Int(hlen))
        headers[shard] = h
        return h
    }

    func fd(_ shard: String) -> Int32 {
        if let f = fds[shard] { return f }
        let f = open(dir.appendingPathComponent(shard).path, O_RDONLY)
        fds[shard] = f
        return f
    }

    /// (layer,proj,part) の 1-expert スライスのバイト数（= 全 layer 共通の slot サイズ）。
    public func sliceBytes(_ layer: Int, _ proj: String, _ part: String) throws -> Int {
        let name = key(layer, proj, part)
        guard let shard = wm[name], let t = try header(shard).meta[name] else {
            throw NSError(domain: "ExpertSource", code: 2)
        }
        return (t.end - t.begin) / t.shape[0]
    }

    public func restShape(_ layer: Int, _ proj: String, _ part: String) throws -> [Int] {
        let name = key(layer, proj, part)
        guard let shard = wm[name], let t = try header(shard).meta[name] else {
            throw NSError(domain: "ExpertSource", code: 2)
        }
        return Array(t.shape.dropFirst())
    }

    public func partDType(_ layer: Int, _ proj: String, _ part: String) throws -> DType {
        let name = key(layer, proj, part)
        guard let shard = wm[name], let t = try header(shard).meta[name] else {
            throw NSError(domain: "ExpertSource", code: 2)
        }
        return ExpertSource.dtype(t.dtype)
    }

    /// expert e の (layer,proj,part) バイトを buf に直接 pread（arena slot へのゼロコピー書込）。
    public func preadInto(_ buf: UnsafeMutableRawPointer, _ layer: Int, _ proj: String,
                          _ part: String, _ e: Int) throws {
        let name = key(layer, proj, part)
        guard let shard = wm[name], let t = try header(shard).meta[name] else {
            throw NSError(domain: "ExpertSource", code: 2)
        }
        let (_, dataStart) = try header(shard)
        let stride = (t.end - t.begin) / t.shape[0]
        let offset = dataStart + t.begin + e * stride
        let n = pread(fd(shard), buf, stride, off_t(offset))
        if n != stride { throw NSError(domain: "ExpertSource", code: 3) }
    }

    /// expert e の (layer,proj,part) スライスを [1, rest...] の MLXArray で返す（pread, 自前 buffer 所有）。
    public func slice(_ layer: Int, _ proj: String, _ part: String, _ e: Int) throws -> MLXArray {
        let name = key(layer, proj, part)
        guard let shard = wm[name] else { throw NSError(domain: "ExpertSource", code: 1) }
        let (meta, dataStart) = try header(shard)
        guard let t = meta[name] else { throw NSError(domain: "ExpertSource", code: 2) }
        let nExp = t.shape[0]
        let stride = (t.end - t.begin) / nExp
        let offset = dataStart + t.begin + e * stride
        let buf = UnsafeMutableRawPointer.allocate(byteCount: stride, alignment: 16)
        let n = pread(fd(shard), buf, stride, off_t(offset))
        if n != stride { buf.deallocate(); throw NSError(domain: "ExpertSource", code: 3) }
        let restShape = [1] + Array(t.shape.dropFirst())
        return MLXArray(rawPointer: buf, restShape, dtype: ExpertSource.dtype(t.dtype)) {
            buf.deallocate()
        }
    }
}

public enum ExpertSourceValidation {
    public static func run(modelDir: String) throws -> String {
        let store = try WeightStore(modelDir: modelDir)
        store.residentNonExperts()
        let src = try ExpertSource(modelDir: modelDir)
        // layer 0,3,39 × proj × part × expert {0,7,255} を resident と bit 比較
        var worst: Float = 0
        var checks = 0
        for layer in [0, 3, 39] {
            for proj in ExpertSource.projs {
                for part in ExpertSource.parts {
                    let full = store.req("\(ExpertSource.prefix).\(layer).mlp.switch_mlp.\(proj).\(part)")
                    for e in [0, 7, 255] {
                        let s = try src.slice(layer, proj, part, e)
                        let ref = full[e ..< (e + 1)]
                        // uint32 は ==、float は abs 差
                        let d: Float
                        if s.dtype == .uint32 {
                            d = MLX.sum(MLX.notEqual(s, ref)).item(Int.self) > 0 ? 1 : 0
                        } else {
                            d = MLX.max(MLX.abs(s.asType(.float32) - ref.asType(.float32))).item(Float.self)
                        }
                        worst = max(worst, d); checks += 1
                    }
                }
            }
        }
        let ok = worst == 0
        return String(format: "[S1] ExpertSource pread スライス: %d 件検証 worst|Δ|=%.3e  %@",
                      checks, worst, ok ? "OK ✅ resident と bit一致" : "MISMATCH ❌")
    }
}
